janus_server/src/aggregator.rs

```diff
@@ -439,11 +442,163 @@ where
+        transitions: Vec<Transition>,
     ) -> Result<AggregateResp, Error> {
-        todo!() // TODO(brandon): implement
+        let vdaf = Arc::new(self.vdaf.clone());
+        let verify_param = Arc::new(self.verify_param.clone());
+        let transitions = Arc::new(transitions);
+
+        Ok(self
+            .datastore
+            .run_tx(|tx| {
+                let vdaf = vdaf.clone();
+                let verify_param = verify_param.clone();
+                let transitions = transitions.clone();
+
+                Box::pin(async move {
+                    // Read existing state.
+                    let (mut aggregation_job, report_aggregations) = future::try_join(
+                        tx.get_aggregation_job::<A>(task_id, job_id),
+                        tx.get_report_aggregations_for_aggregation_job::<A>(
+                            &verify_param,
+                            task_id,
+                            job_id,
+                        ),
+                    )
+                    .await?;
+
+                    // Handle each transition in the request.
+                    let mut report_aggregations = report_aggregations.into_iter();
+                    let (mut saw_continue, mut saw_finish) = (false, false);
+                    let mut response_transitions = Vec::new();
+                    for transition in transitions.iter() {
+                        // Match transition received from leader to stored report aggregation, and
+                        // extract the stored preparation step.
+                        let mut report_aggregation = loop {
+                            let mut report_agg = report_aggregations.next().ok_or_else(|| {
+                                warn!(?task_id, ?job_id, nonce = %transition.nonce, "Leader sent unexpected or out-of-order transitions");
+                                datastore::Error::User(Box::new(Error::UnrecognizedMessage(
+                                    "leader sent unexpected or out-of-order transitions",
+                                    task_id,
+                                )))
+                            })?;
+                            if report_agg.nonce != transition.nonce {
+                                // TODO(brandon): assuming we aren't going to fail due to
+                                // out-of-order/unexpected transitions, any report aggregations that
+                                // get skipped here are skipped because the leader omitted them
+                                // because something failed on the leader's side. We move to INVALID
+                                // state to ensure we don't get later updates, but is that the
+                                // desired behavior?
+                                report_agg.state = ReportAggregationState::Invalid;
+                                tx.update_report_aggregation(&report_agg).await?;
+                                continue;
```

Comment:
I'm not sure this is right. The [relevant protocol text](https://github.com/abetterinternet/ppm-specification/blob/b734ac2272e58b839b8c5d5511e518e42b324956/draft-gpew-priv-ppm.md?plain=1#L1211), from the interop branch:

```
Next, the helper processes the Transition messages from the leader. If any
message appears out of order or has an unrecognized nonce, or if any two
messages have the same nonce, then the helper MUST abort with error
"unrecognizedMessage".
```

My reading of this code is that we'll iterate over the received transitions, and for each one, we'll then iterate over the stored aggregations until we find one whose nonce matches the transition, skipping any stored reports that are unrecognized or in the wrong order. But my reading of the protocol text is that if the leader's i-th transition doesn't match our i-th `report_aggregation`, we should abort the entire `AggregateContinue`.

So I think all this can be simplified if you do something like:
```
for (transition, report_aggregation) in transitions.into_iter().zip(report_aggregations.into_iter()) {
    if transition.nonce != report_aggregation.nonce {
        return Err(/* Some error that aborts the whole request */
    }
    // handle the aggregate continue
}
```

Reply:
By my reading, I think the helper does need to handle cases where related aggregate continue requests have fewer transitions than preceding aggregate init/aggregate continue requests. We do need to be more liberal than zipping the iterators and comparing nonces, and the inner loop that advances `report_aggregations` looks right.

The end of the section "Aggregate Continuation Request" says "The leader processes this message as described in {{aggregate-init-request}}." That section says, at the end of request-response processing, "Finally, the leader removes from the report sequence any report for which the corresponding state is FAILED and proceeds as described in the next section, even if the sequence is empty." The prep-leader diagram says that if the helper failed, the leader's state for the report will be FAILED, but there is also an edge from the leader's invocation of prep_next to FAILED. (with the caveat that the edge label says `prep_next` `send Transition`) Thus, in between making requests to the helper's `/aggregate`, the leader can remove reports from the sequence for two reasons, either the helper's `prep_next()` failed, and it sent a transition of type `failed` to the leader, or the leader's `prep_next()` failed, and by my reading, the leader never sends a transition of type FAILED in a request.

As for the related question in the TODO comment of what we should do with the state enum: If we take the state machine literally, if a nonce stops showing up in requests, then it would stay at the WAITING state indefinitely. On the other hand, we can infer that either the leader's `prep_next()` failed, or the leader is not following the protocol correctly. As FAILED is a terminal state, I think it make sense to store that. If the leader operates correctly, then the helper will never look up this state again. If the leader errs by excluding a report from one continue request, then including it in another request, storing a FAILED state signifies that we don't expect to see the report again, so we would filter out the report when looking up report aggregations, and then correctly reply with `unrecognizedMessage` due to the mismatch.

Reply:
^ Dave's response above is why I implemented it this way. A simple zip doesn't work because, in general, the leader will be sending transitions for a _subsequence_ of reports included in the aggregation job. I can't do something simple based on `Iterator::find`, either, because I want to update the report aggregations that end up being skipped.

The existing logic would also catch duplicate nonces in the leader's request; I updated the error message to note this.

I moved to a terminal state because I wanted to notice/fail out on the "leader omits a report from one continue request, then includes it on a later continue request" case. I think `FAILED` makes a lot of sense (over `INVALID`), but the one wrinkle is that I'm not sure what error code we would want to use -- none of them seem appropriate, and the leader doesn't tell us why it omitted the request (nor do I think we necessarily want to make it do so). `report-dropped` is the closest choice of the existing errors IMO, but currently it specifically indicates that a report was dropped from storage for being "too old" & I'm wary of overloading codes. I'm not sure if introducing a new code is appropriate without a spec change -- `TransitionError` is defined in the spec. Thoughts on the correct error code to use, if we want to switch from `INVALID` to `FAILED`?

Reply:
BTW: for future (probably post-interop-test, unless the CF folks think otherwise), I think we should probably drop the ordering requirement entirely.

The justification I've heard for the ordering requirement is that it saves us from having to do an `O(n lg n)` sort to match incoming transitions to stored report aggregation state. But even without the ordering requirement we could just put the report aggregations in e.g. a `HashMap<Nonce, ReportAggregation>` (in `O(n)` time), then look them up by nonce in `O(1)` time -- no `O(n lg n)`.

The downside of the ordering semantics is the additional implementation complexity, and additional failure modes.

Am I missing something? If not, I'll put this PR on my to-do list.

Reply:
I buy the argument of the leader filtering out `FAILED` reports. I tend to agree with you about removing the ordering requirement, or I had suggested always requiring ascending nonces within an aggregation job so that we get a canonical representation on the wire.

Reply:
I think `unrecognizedMessage` doesn't make sense here -- that's an `aggregator::Error` / what would be reflected in a problem document / an error at the level of an entire request.

Since we only want to fail this individual client report, not the entire request, we need to choose an error from [`TransitionError`](https://github.com/divviup/janus/blob/2c053feb6bfd01db8fd3c7c3c5024132262df549/janus_server/src/message.rs#L762-L773). That is, we want to choose from among the values: `BatchCollected`, `ReportReplayed`, `ReportDropped`, `HpkeUnknownConfigId`, `HpkeDecryptError`, `VdafPrepError`, `UnrecognizedNonce` (or standardize a new value).

Reply:
For now, I went with `ReportDropped`. I'm a little unhappy with the overloading of this error code, but I suppose it's the best one available (and I don't necessarily want to try to specify a new one).

Reply:
I also decided to only update the state if we aren't already in a non-WAITING state, to avoid potentially overwriting a previously-recorded error.